src/loader/mod.rs

```diff
@@ -0,0 +1,580 @@
+        }
+
+        // Sanity checks
+        if ehdr.e_ident[elf::EI_MAG0 as usize] != elf::ELFMAG0 as u8
+            || ehdr.e_ident[elf::EI_MAG1 as usize] != elf::ELFMAG1
+            || ehdr.e_ident[elf::EI_MAG2 as usize] != elf::ELFMAG2
+            || ehdr.e_ident[elf::EI_MAG3 as usize] != elf::ELFMAG3
+        {
+            return Err(Error::InvalidElfMagicNumber);
+        }
+        if ehdr.e_ident[elf::EI_DATA as usize] != elf::ELFDATA2LSB as u8 {
+            return Err(Error::BigEndianElfOnLittle);
+        }
+        if ehdr.e_phentsize as usize != mem::size_of::<elf::Elf64_Phdr>() {
+            return Err(Error::InvalidProgramHeaderSize);
+        }
+        if (ehdr.e_phoff as usize) < mem::size_of::<elf::Elf64_Ehdr>() {
+            // If the program header is backwards, bail.
+            return Err(Error::InvalidProgramHeaderOffset);
+        }
+        if (lowest_kernel_start.is_some())
+            && ((ehdr.e_entry as u64) < lowest_kernel_start.unwrap().raw_value())
+        {
+            return Err(Error::InvalidEntryAddress);
+        }
+
+        let mut loader_result: KernelLoaderResult = Default::default();
+        // where the kernel will be start loaded.
+        loader_result.kernel_load = match kernel_start {
+            Some(start) => GuestAddress(start.raw_value() + (ehdr.e_entry as u64)),
+            None => GuestAddress(ehdr.e_entry as u64),
+        };
+
+        kernel_image
+            .seek(SeekFrom::Start(ehdr.e_phoff))
+            .map_err(|_| Error::SeekProgramHeader)?;
+        let phdrs: Vec<elf::Elf64_Phdr> = unsafe {
+            // Reading the structs is safe for a slice of POD structs.
+            struct_util::read_struct_slice(kernel_image, ehdr.e_phnum as usize)
+                .map_err(|_| Error::ReadProgramHeader)?
+        };
+
+        // Read in each section pointed to by the program headers.
+        for phdr in &phdrs {
+            if phdr.p_type != elf::PT_LOAD || phdr.p_filesz == 0 {
+                continue;
+            }
+
+            kernel_image
+                .seek(SeekFrom::Start(phdr.p_offset))
+                .map_err(|_| Error::SeekKernelStart)?;
+
+            // vmm does not specify where the kernel should be loaded, just
+            // load it to the physical address p_paddr for each segment.
+            let mem_offset = match kernel_start {
+                Some(start) => start
+                    .checked_add(phdr.p_paddr as u64)
+                    .ok_or(Error::InvalidProgramHeaderAddress)?,
+                None => GuestAddress(phdr.p_paddr as u64),
+            };
+
+            guest_mem
+                .read_exact_from(mem_offset, kernel_image, phdr.p_filesz as usize)
+                .map_err(|_| Error::ReadKernelImage)?;
+
+            loader_result.kernel_end =
+                mem_offset.raw_value() as GuestUsize + phdr.p_memsz as GuestUsize;
+        }
+
+        loader_result.setup_header = None;
```

Comment:
Please explain why the setup_header is None in this case.

Reply:
Add comment in code here.